keps/sig-network/4427-relaxed-dns-search-validation/README.md

```diff
@@ -0,0 +1,484 @@
+
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  namespace: default
+  name: dns-example
+spec:
+  containers:
+    - name: test
+      image: nginx
+  dnsPolicy: "None"
+  dnsConfig:
+    nameservers:
+      - 1.2.3.4
+    searches:
+      - abc_d.example.com
+```
+
+However, this returns an error:
+
+```
+The Pod "dns-example" is invalid: spec.dnsConfig.searches[0]: Invalid value: "abc_d.example.com": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
+```
+
+Allowing underscores in the search string allows integration with legacy workloads without allowing anyone to define
+these names within Kubernetes. Since having underscores in a name creates other issues (such as inability to obtain a publicly trusted TLS certificate),
+search strings seem like the only area where this is likely to occur.
+
+### Goals
+
+- Support workloads that need to resolve DNS short names where the full DNS name includes an underscore (`_`).
+
+## Proposal
+
+Introduce a RelaxedDNSSearchValidation feature gate which is disabled by default. When the feature gate is enabled,
+a new DNS name validation function will be used, which keeps the existing validation logic but also allows an underscore (`_`) in any place
+where a dash (`-`) would be allowed currently.
+
+Since the relaxed check allows previously invalid values, care must be taken to support cluster downgrades safely. To accomplish this, the validation will distinguish between new resources and updates to existing resources:
+- When the feature gate is disabled:
+  - (a) New resources will use strict validation based on RFC-1123 (no change to current validation)
+  - (b) Updates to existing resources will use relaxed validation if any search string in the existing list fails strict validation
+- When the feature gate is enabled:
+  - (c) New resources will use relaxed validation.
+  - (d) Updates to existing resources will use relaxed validation.
+
+This means that it is safe to downgrade a cluster with the feature gate enabled to a version where the feature gate is present (whether it’s enabled or disabled). It is not safe, in general, to downgrade from a cluster with the gate enabled to a version prior to the gate being introduced, since values may have been written to storage which will no longer pass validation. However, this scenario requires opting in through enabling the gate. In practice, the recommended approach would be to only enable to the gate after upgrading from a version with relaxed checking already present.
+
+As long as the gate is disabled, there is no compatibility change, so cluster downgrades are not affected by the feature.
+
+### Risks and Mitigations
+
+The change is opt-in, since it requires configuring a search string with an underscore. So there is no risk beyond
```

Comment:
any libc or muslc possible problems?

Reply:
Not beyond what already exists (which is that different implementations of libc may have different length requirements). I checked glibc and muslc and they don't validate anything besides the length when parsing resolv.conf (same is true of golang's resolver).